### ISSUE TITLE
Incorrect result with UNION ALL with constant subquery

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#9874]
### Greenplum version or build
master. Works on 6X_STABLE and 5X_STABLE.

### Expected behavior

```

postgres=# create temp table p (a int);
psql: NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
postgres=# select *                    
from (
      select * from p
      union all
      select 1
     ) s(a)
where s.a = 1;
 a 
---
 1
(1 row)

```

### Actual behavior

It returns no rows:
```
 a 
---
(0 rows)
```

The problem is seen in the plan:
```
postgres=# explain select *
from (
      select * from p
      union all
      select 1
     ) s(a)
where s.a = 1;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..1307.65 rows=98 width=4)
   ->  Append  (cost=0.00..1304.73 rows=98 width=4)
         ->  Seq Scan on p  (cost=0.00..1303.75 rows=97 width=4)
               Filter: (a = 1)
         ->  Result  (cost=0.00..0.03 rows=1 width=4)
               One-Time Filter: (gp_execution_segment() = 2)
               ->  Result  (cost=0.00..0.01 rows=1 width=4)
 Optimizer: Postgres query optimizer
(8 rows)
```

The Result produces a row only on segment 2, but the thanks to direct dispatch, the query is only dispatched to one segment, and that segment is not 2. Commit 75c4769a00e8ab28015651525503f516536a049f added the support for generating plans like the above, with gp_execution_segment(), but it is interacting badly with direct dispatch. It works if you `set gp_enable_direct_dispatch = off;`.

**Comment by kainwen** [comment]
I have spent some time on this issue. The following is my findings:

* ORCA generate a similar plan using the same `Result plannode` skill, but ORCA does not use direct dispatch, so ORCA's plan is correct.

```sql
gpadmin=# set  optimizer = on;
SET
gpadmin=# explain select *
from (
      select * from p
      union all
      select 1
     ) s(a)
where s.a = 1;
                                  QUERY PLAN
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=2 width=4)
   ->  Append  (cost=0.00..431.00 rows=1 width=4)
         ->  Seq Scan on p  (cost=0.00..431.00 rows=1 width=4)
               Filter: (a = 1)
         ->  Result  (cost=0.00..0.00 rows=1 width=4)
               One-Time Filter: (gp_execution_segment() = 0)
               ->  Result  (cost=0.00..0.00 rows=1 width=4)
                     Filter: ((1) = 1)
                     ->  Result  (cost=0.00..0.00 rows=1 width=1)
 Optimizer: Pivotal Optimizer (GPORCA)
```

* Planner will invoke `DirectDispatchUpdateContentIdsFromPlan` to determine the directdispatch info for a slice, and
   the function `DirectDispatchUpdateContentIdsFromPlan`  is only invoked at 3 places:
   1. `create_scan_plan`
   2. `create_motion_plan`
   3. `create_bitmap_subplan`

* For this issue, when generating the plan of append, we recursively generate each subplan:
   1. when generating subplan `Seq Scan on p`, we will invoke `DirectDispatchUpdateContentIdsFromPlan` to update the direct dispatch info in `root->curSlice->directDispatch`.
   2. when generating subplan of `Result`, we have no code path to invoke `DirectDispatchUpdateContentIdsFromPlan` so `root->curSlice->directDispatch` is not modified at this stage.
   3. However, in the above step 2, the subplan can be directly dispatched but with different contentID from step 1
   4. the whole plan of append will be direct dispatched totally based on the info in step1.


------------------------------

So based on all the above analysis,  this bug should happen in a wider scope:

* when planner is generating append plan
* one of the subplans of append plan can be direct dispatch and during planning it, we invoke `DirectDispatchUpdateContentIdsFromPlan`
* another one of the subplans of append plan cannot be (or can only be direct dispatched to a different content from the above one), but when generating this subplan, we do not invoke `DirectDispatchUpdateContentIdsFromPlan`

The issue of this story is a case for the above condition.  Until now, I have not come up with another bad case for this.

**Comment by kainwen** [comment]
Now I have spent some time on debugging the issue. Every plan(or subplan)'s leaf should be a scan.

And invoke directdispatch fuction in create_scan_plan, create_motion_plan, create_bitmap_subplan should be enough.

But when we adopting the Result plannode skills to turn a general locus into partitioned locus, we have to reconsider direct dispatch things.

This issue use value scan to make a general locus subplan, and for values scan, we just skip in the function of `DirectDispatchUpdateContentIdsFromPlan` and we
do nothing when create_projection_plan using the skill above. This is the root cause of this issue.

-----------------------

And the same skill is used also for replicated table or function scan, but it is OK with dd:

```sql
gpadmin=# explain select * from t where a = 1 union all select * from trep;
                                    QUERY PLAN
-----------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..4295.64 rows=96397 width=4)
   ->  Append  (cost=0.00..2367.71 rows=32133 width=4)
         ->  Seq Scan on t  (cost=0.00..1303.75 rows=33 width=4)
               Filter: (a = 1)
         ->  Result  (cost=0.00..2026.00 rows=32100 width=4)
               One-Time Filter: (gp_execution_segment() = 1)
               ->  Seq Scan on trep  (cost=0.00..1063.00 rows=32100 width=4)
 Optimizer: Postgres query optimizer

gpadmin=# explain select * from t where a = 1 union all select * from generate_series(1, 10);
                                       QUERY PLAN
-----------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1336.64 rows=1097 width=4)
   ->  Append  (cost=0.00..1314.72 rows=366 width=4)
         ->  Seq Scan on t  (cost=0.00..1303.75 rows=33 width=4)
               Filter: (a = 1)
         ->  Result  (cost=0.00..20.00 rows=334 width=4)
               One-Time Filter: (gp_execution_segment() = 1)
               ->  Function Scan on generate_series  (cost=0.00..10.00 rows=334 width=4)
 Optimizer: Postgres query optimizer
``` 

This is because function_scan and table scan(not partitioned policy) will set directdispatch to false.

--------------------

Based on the above analysis, I think a reasonable fix is:

in the function create_projection_plan, if we use the `gp_execution_segment()` skill, let's update directdispatch info.

**Comment by kainwen** [comment]
Fixed by https://github.com/greenplum-db/gpdb/pull/9948 .

**Comment by hlinnaka** [comment]
> Fixed by #9948 .

Thanks!
